crates/arroyo-df/src/udafs.rs

```diff
@@ -0,0 +1,228 @@
+use crate::physical::UdfDylib;
+use arrow::buffer::OffsetBuffer;
+use arrow_array::cast::as_list_array;
+use arrow_array::{new_empty_array, Array, ListArray};
+use arrow_schema::{DataType, FieldRef, IntervalUnit, TimeUnit};
+use datafusion::arrow::array::ArrayRef;
+use datafusion::scalar::ScalarValue;
+use datafusion::{error::Result, physical_plan::Accumulator};
+use datafusion_common::DataFusionError;
+use datafusion_expr::{ColumnarValue, ScalarUDFImpl};
+use std::fmt::Debug;
+use std::sync::Arc;
+
+#[derive(Debug, Clone)]
+pub struct UdafArg {
+    values: Vec<ArrayRef>,
+    inner: FieldRef,
+}
+
+impl UdafArg {
+    pub fn new(inner: FieldRef) -> Self {
+        UdafArg {
+            values: vec![],
+            inner,
+        }
+    }
+
+    fn concatenate_array(&self) -> Result<ListArray> {
+        let element_arrays: Vec<&dyn Array> = self.values.iter().map(|a| a.as_ref()).collect();
+
+        let arr = arrow::compute::concat(&element_arrays)?;
+
+        Ok(list_from_arr(&self.inner, arr))
+    }
+}
+
+/// An Arroyo UDAF is a scalar function that takes vector arguments. This Accumulator infra
+/// exists to wrap an Arroyo UDAF in a DF UDAF that first accumulates the array of data, then
+/// passes it to the vector-taking UDF.
+#[derive(Debug)]
+pub struct ArroyoUdaf {
+    args: Vec<UdafArg>,
+    output_type: Arc<DataType>,
+    udf: UdfDylib,
+}
+
+impl ArroyoUdaf {
+    pub fn new(args: Vec<UdafArg>, output_type: Arc<DataType>, udf: UdfDylib) -> Self {
+        assert!(
+            args.len() > 0,
+            "UDAF {} has no arguments, but UDAFs must have at least one",
+            udf.name()
+        );
+        ArroyoUdaf {
+            args,
+            output_type,
+            udf,
+        }
+    }
+}
+
+impl Accumulator for ArroyoUdaf {
+    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
+        if values.is_empty() {
+            return Ok(());
+        }
+
+        for (arg, v) in self.args.iter_mut().zip(values) {
+            arg.values.push(v.clone());
+        }
+
+        Ok(())
+    }
+
+    fn evaluate(&mut self) -> Result<ScalarValue> {
+        if self.args[0].values.is_empty() {
+            return Ok(scalar_none(&self.output_type));
+        }
+
+        let args: Result<Vec<_>> = self
+            .args
+            .iter()
+            .map(|arg| {
+                Ok(ColumnarValue::Scalar(ScalarValue::List(Arc::new(
+                    arg.concatenate_array()?,
+                ))))
+            })
+            .collect();
+
+        let ColumnarValue::Scalar(scalar) = self.udf.invoke(&args?[..]).unwrap() else {
+            return Err(DataFusionError::Execution(format!(
+                "UDAF {} returned an array result",
+                self.udf.name()
+            )));
+        };
+
+        Ok(scalar)
+    }
+
+    fn size(&self) -> usize {
+        std::mem::size_of_val(self)
```

Comment:
This is supposed to be the total size of the allocator, i.e. include the sizes of the arrays it owns.